File: editor/Assets/Haptic.cs
//
// Haptic.cs by Xein
// 04 Jul 2026
//

using System;
using System.IO;
using Tomlyn;
using Tomlyn.Model;

namespace editor.Assets;

public enum HapticMode {
	Standard,
	Curve,
	AdaptiveTrigger,
	AudioHaptic
}

public enum HapticChannels {
	Single,
	Dual
}

public sealed class Haptic {
	public string SchemaUid { get; set; } = "";
	public HapticMode Mode { get; set; } = HapticMode.Standard;
	public int Priority { get; set; }
	public int DurationMs { get; set; } = 200;

	// standard mode
	public float Left { get; set; } = 0.5f;
	public float Right { get; set; } = 0.5f;

	// curve mode
	public HapticChannels Channels { get; set; } = HapticChannels.Single;
	public float Pan { get; set; }
	public float Multiplier { get; set; } = 1f;
	public Curve? Curve { get; set; }
	public Curve? CurveRight { get; set; }

	public static Haptic FromFile(string path) {
		return FromString(File.ReadAllText(path));
	}

	public static Haptic FromString(string toml) {
		var table = TomlSerializer.Deserialize<TomlTable>(toml)
			?? throw new FormatException("Haptic: failed to parse TOML");

		var h = new Haptic {
			SchemaUid = GetString(table, "schema", ""),
			Mode = ParseMode(GetString(table, "mode", "standard")),
			Priority = GetInt(table, "priority", 0),
			DurationMs = GetInt(table, "duration_ms", 200),
			Left = GetFloat(table, "left", 0.5f),
			Right = GetFloat(table, "right", 0.5f),
			Channels = GetString(table, "channels", "single") == "dual"
				? HapticChannels.Dual
				: HapticChannels.Single,
			Pan = GetFloat(table, "pan", 0f),
			Multiplier = GetFloat(table, "multiplier", 1f)
		};

		if (table.TryGetValue("curve", out var c) && c is TomlTable ct)
			h.Curve = Curve.FromToml(ct);
		if (table.TryGetValue("curve_right", out var cr) && cr is TomlTable crt)
			h.CurveRight = Curve.FromToml(crt);

		return h;
	}

	public void Save(string path) {
		File.WriteAllText(path, Serialize());
	}

	public string Serialize() {
		var table = new TomlTable();
		if (SchemaUid.Length > 0)
			table["schema"] = SchemaUid;
		table["mode"] = ModeToString(Mode);
		table["priority"] = (long)Priority;
		table["duration_ms"] = (long)DurationMs;

		switch (Mode) {
			case HapticMode.Standard:
				table["left"] = (double)Left;
				table["right"] = (double)Right;
				break;
			case HapticMode.Curve:
				table["channels"] = Channels == HapticChannels.Dual ? "dual" : "single";
				table["pan"] = (double)Pan;
				table["multiplier"] = (double)Multiplier;
				if (Curve is { } curve)
					table["curve"] = curve.ToToml();
				if (Channels == HapticChannels.Dual && CurveRight is { } curveRight)
					table["curve_right"] = curveRight.ToToml();
				break;
		}

		return TomlSerializer.Serialize(table);
	}

	private static string ModeToString(HapticMode m) {
		return m switch {
			HapticMode.Standard => "standard",
			HapticMode.Curve => "curve",
			HapticMode.AdaptiveTrigger => "adaptive_trigger",
			HapticMode.AudioHaptic => "audio_haptic",
			_ => "standard"
		};
	}

	private static HapticMode ParseMode(string s) {
		return s switch {
			"curve" => HapticMode.Curve,
			"adaptive_trigger" => HapticMode.AdaptiveTrigger,
			"audio_haptic" => HapticMode.AudioHaptic,
			_ => HapticMode.Standard
		};
	}

	private static string GetString(TomlTable t, string key, string fallback) {
		return t.TryGetValue(key, out var v) && v is string s ? s : fallback;
	}

	private static int GetInt(TomlTable t, string key, int fallback) {
		return t.TryGetValue(key, out var v)
			? v switch { long l => (int)l, int i => i, double d => (int)d, _ => fallback }
			: fallback;
	}

	private static float GetFloat(TomlTable t, string key, float fallback) {
		return t.TryGetValue(key, out var v)
			? v switch { double d => (float)d, float f => f, long l => l, int i => i, _ => fallback }
			: fallback;
	}
}

File: engine/src/toast/input/assets/haptic.cpp
#include "haptic.hpp"

#include <algorithm>
#include <stdexcept>
#include <string>
#include <toast/log.hpp>

namespace assets {

namespace {

auto hapticModeFromString(std::string_view s) -> HapticMode {
	if (s == "curve") {
		return HapticMode::curve;
	}
	if (s == "adaptive_trigger") {
		return HapticMode::adaptive_trigger;
	}
	if (s == "audio_haptic") {
		return HapticMode::audio_haptic;
	}
	return HapticMode::standard;
}

}

Haptic::Haptic(const toml::table& table, Handle<Schema> schema) : Data(table, std::move(schema), Data::keep_all_keys) {
	const auto& d = static_cast<const DataValue&>(m_root);

	std::string mode_str;
	if (d.contains("mode")) {
		mode_str = d["mode"].as<std::string>();
	}
	m_mode = hapticModeFromString(mode_str);

	if (d.contains("priority")) {
		m_priority = static_cast<int>(d["priority"].as<int64_t>());
	}
	if (d.contains("duration_ms")) {
		m_duration_ms = static_cast<uint32_t>(d["duration_ms"].as<int64_t>());
	}

	switch (m_mode) {
		case HapticMode::standard:
			if (d.contains("left")) {
				m_left = static_cast<float>(d["left"].as<double>());
			}
			if (d.contains("right")) {
				m_right = static_cast<float>(d["right"].as<double>());
			}
			break;

		case HapticMode::curve: {
			std::string channels_str;
			if (d.contains("channels")) {
				channels_str = d["channels"].as<std::string>();
			}
			m_channels = (channels_str == "dual") ? HapticChannels::dual : HapticChannels::single;

			if (d.contains("pan")) {
				m_pan = std::clamp(static_cast<float>(d["pan"].as<double>()), -1.0f, 1.0f);
			}
			if (d.contains("multiplier")) {
				m_multiplier = static_cast<float>(d["multiplier"].as<double>());
			}

			if (d.contains("curve") && d["curve"].isObject()) {
				try {
					m_curve = Curve::fromToml(d["curve"].asTomlTable());
				} catch (const std::exception& e) { TOAST_WARN("Haptic", "Failed to parse [curve]: {}", e.what()); }
			}
			if (!m_curve) {
				TOAST_WARN("Haptic", "curve mode is missing a valid [curve] table");
			}

			if (m_channels == HapticChannels::dual) {
				if (d.contains("curve_right") && d["curve_right"].isObject()) {
					try {
						m_curve_right = Curve::fromToml(d["curve_right"].asTomlTable());
					} catch (const std::exception& e) { TOAST_WARN("Haptic", "Failed to parse [curve_right]: {}", e.what()); }
				}
				if (!m_curve_right) {
					TOAST_WARN("Haptic", "dual channels need a valid [curve_right] table");
				}
			}
			break;
		}

		case HapticMode::adaptive_trigger:
		case HapticMode::audio_haptic: break;
	}
}

auto Haptic::sample(float t01) const -> glm::vec2 {
	const float t = std::clamp(t01, 0.0f, 1.0f);

	switch (m_mode) {
		case HapticMode::standard: return {std::clamp(m_left, 0.0f, 1.0f), std::clamp(m_right, 0.0f, 1.0f)};

		case HapticMode::curve: {
			if (!m_curve) {
				return {0.0f, 0.0f};
			}

			const float left_y = m_curve->evalAtX(t) * m_multiplier;

			if (m_channels == HapticChannels::dual) {
				const float right_y = m_curve_right ? m_curve_right->evalAtX(t) * m_multiplier : 0.0f;
				return {std::clamp(left_y, 0.0f, 1.0f), std::clamp(right_y, 0.0f, 1.0f)};
			}

			// -1 left .. +1 right
			const float left_gain = m_pan > 0.0f ? 1.0f - m_pan : 1.0f;
			const float right_gain = m_pan < 0.0f ? 1.0f + m_pan : 1.0f;
			return {std::clamp(left_y * left_gain, 0.0f, 1.0f), std::clamp(left_y * right_gain, 0.0f, 1.0f)};
		}

		case HapticMode::adaptive_trigger:
		case HapticMode::audio_haptic: return {0.0f, 0.0f};
	}

	return {0.0f, 0.0f};
}

}
